### PR TITLE
balancer: Enforce embedding the SubConn interface in implementations

### PR DESCRIPTION
**Pull request by arjan-bal**
Presently, the doc comment on the `balancer.SubConn` interface mentions that external implementers should embed the `balancer.SubConn` interface, however, this is not enforced at compilation time:
```
// This interface is to be implemented by gRPC. Users should not need their own
// implementation of this interface. For situations like testing, any
// implementations should embed this interface. This allows gRPC to add new
// methods to this interface.
```
This PR adds an unexported method on the interface to add a compile time check for enforcing the requirement.

RELEASE NOTES:
* balancer: An unexported method is added to the `balancer.SubConn` interface to force implementors to embed the interface. This requirement is present in the interface documentation, but wasn't enforced earlier.

### CODE REVIEW

File: balancer/balancer.go [hunk 1]
```diff
@@ -155,6 +155,9 @@ type SubConn interface {
 	// indicate the shutdown operation.  This may be delivered before
 	// in-progress RPCs are complete and the actual connection is closed.
 	Shutdown()
+	// enforceEmbedding is an unexported method to enforce implementers embed
```
Comment (dfawley) [inline balancer/balancer.go:158]: "...to _force_ implementers _to_ embed..."  or "to enforce _that_ implementers embed".

**Review by dfawley — APPROVED** [review]
This is a potential breaking change.  I suspect it would only break tests (doing mocking?), but it's possible to break actual users.  That's OK, we just need to be careful to document the change and also take care when importing the code internally.

File: balancer/balancer.go (hunk 1 above)
Comment (arjan-bal) [reply balancer/balancer.go]: Grammar fixed.

**Comment by arjan-bal** [comment]
> This is a potential breaking change. I suspect it would only break tests (doing mocking?), but it's possible to break actual users. That's OK, we just need to be careful to document the change and also take care when importing the code internally.

@dfawley , I checked the internal usages and found that adding a private method will break code generated using GoMock, e.g: https://github.com/GoogleCloudPlatform/grpc-gcp-go/blob/d2bc599b72aeef2d85143b234ca7d602ae7f4302/grpcgcp/mocks/mock_balancer.go#L116-L119

Users could manually edit the generated mocks to embed `balancer.SubConn`. Do you think it's okay to expect users manually edit their mocks?

**Comment by easwars** [comment]
> I checked the internal usages and found that adding a private method will break code generated using GoMock

Hmm.. at least internally, mocking is not recommended for Go. I'm not sure about the external world though. Do you have any experience using mocking frameworks and know what is more commonly used and whether they generate code differently wherein the generated code embeds the mocked interface? (some google searching didn't yield anything useful)

> Users could manually edit the generated mocks to embed balancer.SubConn. Do you think it's okay to expect users manually edit their mocks?

I guess this is no worse if you didn't use mocking and didn't embed, but chose to implement the methods to satisfy the interface.

File: internal/testutils/balancer.go [hunk 2]
```diff
@@ -32,6 +32,7 @@ import (
 
 // TestSubConn implements the SubConn interface, to be used in tests.
 type TestSubConn struct {
+	balancer.SubConn
```
Comment (easwars) [inline internal/testutils/balancer.go:35]: Funny that we ourselves didn't embed the interface in our tests.

**Comment by dfawley** [comment]
> Do you think it's okay to expect users manually edit their mocks?

Probably not if they're generated, unless they're also checked in, and then...maybe.  How many instances of this exist? Is it feasible to fix all the ones that we're (indirectly) responsible for?

**Review by easwars — APPROVED** [review]

**Review by dfawley — APPROVED** [review]
LGTM, but we need to wait to check this in until all the internal dependencies are resolved.

**Comment by arjan-bal** [comment]
The PR in grpc-gcp-go is merged: https://github.com/GoogleCloudPlatform/grpc-gcp-go/pull/89
grpc-gcp-go may need to do a release before we're clear to merge this PR.

**Pull request merged**